src/modules/src/controller_mellinger.c

```diff
@@ -331,98 +342,100 @@ PARAM_GROUP_START(ctrlMel)
  */
-PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, kR_z, &kR_z)
+PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, kR_z, &g_self.kR_z)
 /**
  * @brief Attitude D-gain (roll and pitch)
  */
-PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, kw_xy, &kw_xy)
+PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, kw_xy, &g_self.kw_xy)
 /**
  * @brief Attitude D-gain (yaw)
  */
-PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, kw_z, &kw_z)
+PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, kw_z, &g_self.kw_z)
 /**
  * @brief Attitude I-gain (roll and pitch)
  */
-PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, ki_m_xy, &ki_m_xy)
+PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, ki_m_xy, &g_self.ki_m_xy)
 /**
  * @brief Attitude I-gain (yaw)
  */
-PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, ki_m_z, &ki_m_z)
+PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, ki_m_z, &g_self.ki_m_z)
 /**
  * @brief Angular velocity D-Gain (roll and pitch)
  */
-PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, kd_omega_rp, &kd_omega_rp)
+PARAM_ADD_CORE(PARAM_FLOAT | PARAM_PERSISTENT, kd_omega_rp, &g_self.kd_omega_rp)
 /**
  * @brief Attitude maximum accumulated error (roll and pitch)
  */
-PARAM_ADD(PARAM_FLOAT | PARAM_PERSISTENT, i_range_m_xy, &i_range_m_xy)
+PARAM_ADD(PARAM_FLOAT | PARAM_PERSISTENT, i_range_m_xy, &g_self.i_range_m_xy)
 /**
  * @brief Attitude maximum accumulated error (yaw)
  */
-PARAM_ADD(PARAM_FLOAT | PARAM_PERSISTENT, i_range_m_z, &i_range_m_z)
+PARAM_ADD(PARAM_FLOAT | PARAM_PERSISTENT, i_range_m_z, &g_self.i_range_m_z)
 PARAM_GROUP_STOP(ctrlMel)
 
 /**
  * Logging variables for the command and reference signals for the
  * Mellinger controller
  */
 LOG_GROUP_START(ctrlMel)
-LOG_ADD(LOG_FLOAT, cmd_thrust, &cmd_thrust)
-LOG_ADD(LOG_FLOAT, cmd_roll, &cmd_roll)
-LOG_ADD(LOG_FLOAT, cmd_pitch, &cmd_pitch)
-LOG_ADD(LOG_FLOAT, cmd_yaw, &cmd_yaw)
-LOG_ADD(LOG_FLOAT, r_roll, &r_roll)
-LOG_ADD(LOG_FLOAT, r_pitch, &r_pitch)
-LOG_ADD(LOG_FLOAT, r_yaw, &r_yaw)
-LOG_ADD(LOG_FLOAT, accelz, &accelz)
-LOG_ADD(LOG_FLOAT, zdx, &z_axis_desired.x)
-LOG_ADD(LOG_FLOAT, zdy, &z_axis_desired.y)
-LOG_ADD(LOG_FLOAT, zdz, &z_axis_desired.z)
-LOG_ADD(LOG_FLOAT, i_err_x, &i_error_x)
-LOG_ADD(LOG_FLOAT, i_err_y, &i_error_y)
-LOG_ADD(LOG_FLOAT, i_err_z, &i_error_z)
+LOG_ADD(LOG_FLOAT, cmd_thrust, &g_self.cmd_thrust)
+LOG_ADD(LOG_FLOAT, cmd_roll, &g_self.cmd_roll)
+LOG_ADD(LOG_FLOAT, cmd_pitch, &g_self.cmd_pitch)
+LOG_ADD(LOG_FLOAT, cmd_yaw, &g_self.cmd_yaw)
+LOG_ADD(LOG_FLOAT, r_roll, &g_self.r_roll)
+LOG_ADD(LOG_FLOAT, r_pitch, &g_self.r_pitch)
+LOG_ADD(LOG_FLOAT, r_yaw, &g_self.r_yaw)
+LOG_ADD(LOG_FLOAT, accelz, &g_self.accelz)
+LOG_ADD(LOG_FLOAT, zdx, &g_self.z_axis_desired.x)
+LOG_ADD(LOG_FLOAT, zdy, &g_self.z_axis_desired.y)
+LOG_ADD(LOG_FLOAT, zdz, &g_self.z_axis_desired.z)
+LOG_ADD(LOG_FLOAT, i_err_x, &g_self.i_error_x)
+LOG_ADD(LOG_FLOAT, i_err_y, &g_self.i_error_y)
+LOG_ADD(LOG_FLOAT, i_err_z, &g_self.i_error_z)
 LOG_GROUP_STOP(ctrlMel)
+
+#endif // CRAZYFLIE_FW
```

Comment:
It is not ideal that we put all this code and logging and variables in the CRAZYFLIE_FW define... Technically the struct exists but it is just not written in ? I had a question about the use of this define in one of my earlier questions.

Reply:
The if/endif preprocessor doesn't need to go around logging/params. In this file, it might not technically be needed at all, since the binding compilation process only generates bindings from a given header file.

Reply:
Ah yeah that is true. Would it be okay for you to remove this define in case it is not necessary? I just want to prevent more ifdefs to be added if necessary as it complicates the code eventually.

Reply:
> Technically the struct exists but it is just not written in ?

In `collision_avoidance.c` the global state struct is defined inside the `#ifdef CRAZYFLIE_FW`. The bindings shouldn't need it at all.

> The if/endif preprocessor doesn't need to go around logging/params.

If they are moved outside the `#ifdef CRAZYFLIE_FW`, then the C compiler for the bindings will need to know about the logging/param system, right? When I was working on `collision_avoidance.c` I assumed we wouldn't want that.